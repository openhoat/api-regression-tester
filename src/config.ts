import { z } from 'zod'
import type { ConfigCustom } from './util/config-builder'

const schema = z.object({
  legacyBaseUrl: z.string().default('http://localhost:3000'),
  actualBaseUrl: z.string().default('http://localhost:3001'),
  logLevel: z
    .enum(['silent', 'fatal', 'error', 'warn', 'info', 'debug', 'trace'])
    .default('warn'),
  timeout: z.number().default(5000),
  modelProvider: z.enum(['openai', 'gemini', 'mistral', 'ollama']),
  model: z.string().optional(),
  googleApiKey: z.string().optional(),
  openaiApiKey: z.string().optional(),
  mistralApiKey: z.string().optional(),
  ollamaBaseUrl: z.string().optional(),
})

export type Config = z.infer<typeof schema>

const custom: ConfigCustom<Config> = {}

export const configSpec = { schema, custom }
