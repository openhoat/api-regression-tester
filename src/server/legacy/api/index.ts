import type { FastifyInstance } from 'fastify'
import { booksRouter } from './books'

export const apiRouter = async (fastify: FastifyInstance) => {
  fastify.register(booksRouter, { prefix: '/books' })
}
