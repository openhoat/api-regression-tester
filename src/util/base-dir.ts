import { relative, resolve } from 'node:path'

const relativeBaseDir = '../..'
const thisScriptDir = __dirname
const currentDir = process.cwd()
const absoluteBaseDir = resolve(thisScriptDir, relativeBaseDir)

export const baseDir = relative(currentDir, absoluteBaseDir)
