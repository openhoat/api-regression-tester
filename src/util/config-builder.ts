import type { ZodType, z } from 'zod'
import { loadEnv } from './env.helper'
import { toPascalCase } from './string.helper'

export type ConfigCustomItem = {
  transform?: z.ZodType<unknown>
  envVarName?: string
}

export type ConfigCustom<T> = Partial<Record<keyof T, ConfigCustomItem>>

export type ConfigSpec<S extends Record<string, ZodType>, T> = {
  schema: z.ZodObject<S>
  custom?: ConfigCustom<T>
}

export type EnvProvider = Record<string, string | undefined>

export const buildConfig = <S extends Record<string, ZodType>, T>(
  { schema, custom }: ConfigSpec<S, T>,
  envProvider: EnvProvider = process.env,
) => {
  if (envProvider === process.env) {
    loadEnv(envProvider.ENVIRONMENT)
  }
  const { VERBOSE } = envProvider
  const verbose = VERBOSE === 'true'
  const configShemaEntries = Object.keys(schema.shape)
  const config = schema.parse(
    configShemaEntries.reduce(
      (acc, name: string) => {
        const spec = custom?.[name as keyof T]
        const envVarName = spec?.envVarName ?? toPascalCase(name)
        const envVarValue = envProvider[envVarName]
        if (envVarValue === undefined) {
          return acc
        }
        const transform = spec?.transform ?? schema.shape[name]
        acc[name] = transform.parse(envVarValue)
        return acc
      },
      {} as Record<string, unknown>,
    ),
  )
  if (verbose) {
    console.debug('Loaded config:', config)
  }
  return config
}
