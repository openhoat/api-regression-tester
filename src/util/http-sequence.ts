import { readFile } from 'node:fs/promises'
import { JSONPath } from 'jsonpath-plus'
import _ from 'lodash'
import { parse } from 'yaml'
import {
  type ApiRequest,
  type ApiResponse,
  type HttpMethod,
  sendRequest,
} from './http.helper'
import type { JsonRecord, JsonValue } from './types'

export type HttpSequenceStep = {
  name: string
  request: {
    method: HttpMethod
    endpoint: string
    basicAuth?: {
      username: string
      password: string
    }
    headers?: Record<string, string>
    queryParams?: Record<string, string>
    payload?: unknown
  }
  transformations?: {
    key: string
    type?: string
    value?: unknown
    args?: string[]
  }[]
  assertions?: { name: string; type: string; args?: string[] }[]
}

export type HttpSequence = {
  name: string
  baseUrl?: string
  steps: HttpSequenceStep[]
  responseMapping: Record<string, string>
}

export type ConversationStep = {
  request: ApiRequest
  response?: ApiResponse
}

export type Conversation = Record<string, ConversationStep>

export type HttpSequenceState<T extends JsonRecord> = {
  result?: T
} & JsonRecord

export type HttpSequenceResult<T extends JsonRecord> = {
  conversation: Conversation
  state: HttpSequenceState<T>
}

export type Asserter = (store: JsonRecord, ...args: unknown[]) => void

export type Asserters = Record<string, Asserter>

export type Transformer = (store: JsonRecord, ...args: unknown[]) => unknown

export type Transformers = Record<string, Transformer>

export type SequenceOptions<T extends JsonRecord> = Partial<{
  conversation: Conversation
  state: HttpSequenceState<T>
  asserters: Asserters
  transformers: Transformers
}>

export type SequenceStepOptions<T extends JsonRecord> = Partial<{
  baseUrl: string
}> &
  SequenceOptions<T>

export const evaluateValue = (
  value: string,
  store: JsonRecord,
  defaultValue: unknown = undefined,
) => {
  if (!value.startsWith('$')) {
    return defaultValue
  }
  const [computedValue] = JSONPath({ path: value, json: store })
  return computedValue
}

export const runSequenceStep = async <T extends JsonRecord>(
  step: HttpSequenceStep,
  options?: SequenceStepOptions<T>,
): Promise<HttpSequenceResult<T>> => {
  const { name, request, transformations, assertions } = step
  const {
    conversation = {} satisfies Conversation,
    state = {} as HttpSequenceState<T>,
    asserters,
    transformers,
  } = options ?? {}
  const store = { conversation, state } as unknown as JsonRecord
  const { endpoint, method, queryParams, headers, payload, basicAuth } = request
  if (headers) {
    for (const [key, value] of Object.entries(headers)) {
      headers[key] = `${evaluateValue(value, store, value)}`
    }
  }
  const apiRequest: ApiRequest = {
    method,
    path: endpoint,
    basicAuth,
    headers,
    params: queryParams,
    body: payload,
  }
  if (!conversation[name]) {
    conversation[name] = { request: apiRequest }
  }
  conversation[name].response = await sendRequest<JsonValue>(apiRequest, {
    baseUrl: options?.baseUrl,
  })
  if (transformations) {
    transformations.forEach(({ key, type, value, args = [] }) => {
      const transformer = type && transformers?.[type]
      if (type && !transformer) {
        throw new Error(`Missing transformer for type "${type}"`)
      }
      const finalValue = transformer
        ? transformer(store, ...args)
        : JSONPath({ path: `${value}`, json: store })
      _.set(state, `result.${key}`, finalValue)
    })
  }
  if (assertions) {
    assertions.forEach(({ type, args = [] }) => {
      const asserter = asserters?.[type]
      if (!asserter) {
        throw new Error(`Missing asserter for type "${type}"`)
      }
      asserter(store, ...args)
    })
  }
  return store as unknown as { conversation: Conversation; state: T }
}

export const runSequence = async <T extends JsonRecord>(
  sequence: HttpSequence,
  options?: SequenceOptions<T>,
): Promise<HttpSequenceResult<T>> => {
  const { baseUrl, steps } = sequence
  const {
    conversation = {} satisfies Conversation,
    state = {} as HttpSequenceState<T>,
    asserters,
    transformers,
  } = options ?? {}
  const store = { conversation, state }
  const promise = steps.reduce(
    async (acc: Promise<HttpSequenceResult<T>>, step) => {
      const { conversation, state } = await acc
      return runSequenceStep(step, {
        baseUrl,
        conversation,
        state,
        asserters,
        transformers,
      })
    },
    Promise.resolve(store),
  )
  await promise
  return store
}

export const loadYamlFile = async <T>(file: string): Promise<T> => {
  const content = await readFile(file, 'utf8')
  return parse(content) satisfies T
}
