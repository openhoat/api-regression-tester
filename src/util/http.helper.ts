import axios from 'axios'

export type HttpMethod = 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH'

export interface ApiRequest {
  method: HttpMethod
  path: string
  basicAuth?: {
    username: string
    password: string
  }
  headers?: Record<string, string>
  body?: unknown
  params?: Record<string, unknown>
}

export interface ApiResponse<T = unknown> {
  status: number
  data: T
}

export const sendRequest = async <T = unknown>(
  req: ApiRequest,
  options?: { baseUrl?: string; timeout?: number },
): Promise<ApiResponse<T>> => {
  const { data, status } = await axios({
    auth: req.basicAuth,
    baseURL: options?.baseUrl,
    method: req.method,
    params: req.params,
    url: req.path,
    headers: req.headers || {},
    data: req.body || null,
    validateStatus: () => true,
    timeout: options?.timeout ?? 5000,
  })
  return { status, data }
}
