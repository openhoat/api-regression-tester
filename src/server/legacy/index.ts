import Fastify from 'fastify'
import type { LevelWithSilentOrString } from 'pino'
import type { Service } from '../../util/service'
import { apiRouter } from './api'

export const createServer = (options?: {
  port?: number
  logLevel?: LevelWithSilentOrString
}): Service => {
  const { port, logLevel } = options ?? {}
  const app = Fastify({
    logger: {
      transport: {
        target: 'pino-pretty',
        options: {
          translateTime: 'HH:MM:ss Z',
          ignore: 'pid,hostname',
        },
      },
      level: logLevel,
    },
  })
  app.register(apiRouter, { prefix: '/api' })
  return {
    start: async () => {
      await app.listen({ port: port ?? 3000, host: '0.0.0.0' })
    },
    stop: async () => {
      await app.close()
    },
  }
}
