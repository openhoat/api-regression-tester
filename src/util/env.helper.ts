import { join } from 'node:path'
import { config as dotenvConfig } from 'dotenv'
import { baseDir } from './base-dir'

export const loadDotenvFile = (envName?: string) => {
  const envFilename = envName ? `.env.${envName}` : '.env'
  dotenvConfig({ path: join(baseDir, envFilename), quiet: true })
  const { VERBOSE } = process.env
  const verbose = VERBOSE === 'true'
  if (verbose) {
    console.debug(`Environment loaded from ${envFilename}`)
  }
}

export const loadEnv = (envName?: string) => {
  loadDotenvFile('local')
  if (envName && envName !== 'local') {
    loadDotenvFile(envName)
  }
  loadDotenvFile()
}
