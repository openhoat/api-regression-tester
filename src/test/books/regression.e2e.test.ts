import { afterAll, beforeAll, describe, expect, test } from '@jest/globals'
import { configSpec } from '../../config'
import { createServer as createActualServer } from '../../server/actual'
import { createServer as createLegacyServer } from '../../server/legacy'
import { buildConfig } from '../../util/config-builder'
import { type ApiResponse, sendRequest } from '../../util/http.helper'
import type { Service } from '../../util/service'
import { mockApis } from './api-mocks'
import { compareResponses } from './matchers/comparison.helper'
import { type ServerRequestTestCase, type Store, testcases } from './testcases'

const { legacyBaseUrl, actualBaseUrl, logLevel, timeout } =
  buildConfig(configSpec)

const { MOCK, LOCAL_API, VERBOSE } = process.env
const verbose = VERBOSE === 'true'
const mock = MOCK === 'true'
const localApi = LOCAL_API === 'true'

const validateAssertions = (name: string, res: ApiResponse) => {
  const { status, data } = res
  const hintPrefix = `${name} `
  const statusHint = `${hintPrefix} status`
  const dataHint = `${hintPrefix} data`
  expect(status).toMatchSnapshot(statusHint)
  if (Array.isArray(data)) {
    const expectedArray = Array(data.length).fill({ id: expect.any(String) })
    expect(data).toMatchSnapshot(expectedArray, dataHint)
  } else {
    expect(res.data).toMatchSnapshot({ id: expect.any(String) }, dataHint)
  }
}

const doServerRequest = async (
  baseUrl: string,
  serverRequest: ServerRequestTestCase,
  store: Store,
) => {
  const { before, request, after } = serverRequest
  if (before) {
    before({ store, request })
  }
  const response = await sendRequest(request, {
    baseUrl,
    timeout,
  })
  serverRequest.response = response
  if (after) {
    after({ store, request, response })
  }
  return response
}

describe('API regression test', () => {
  const stores = {
    // biome-ignore lint/suspicious/noExplicitAny: special case
    legacy: new Map<string, any>(),
    // biome-ignore lint/suspicious/noExplicitAny: special case
    actual: new Map<string, any>(),
  }
  const servers: {
    legacy: Service | null
    actual: Service | null
  } = { legacy: null, actual: null }
  beforeAll(async () => {
    if (localApi) {
      const legacyServer = createLegacyServer({ logLevel })
      const actualServer = createActualServer({ logLevel })
      await Promise.all([legacyServer.start(), actualServer.start()])
      servers.legacy = legacyServer
      servers.actual = actualServer
    }
    if (mock) {
      mockApis(legacyBaseUrl, actualBaseUrl)
    }
  })
  afterAll(async () => {
    const { legacy: legacyServer, actual: actualServer } = servers
    if (legacyServer) {
      await legacyServer.stop()
      servers.legacy = null
    }
    if (actualServer) {
      await actualServer.stop()
      servers.actual = null
    }
  })
  test.each(testcases)(
    'should return equivalent results for "$name" request',
    async ({ name, legacy, actual }) => {
      // When
      const [legacyRes, actualRes] = await Promise.all([
        doServerRequest(legacyBaseUrl, legacy, stores.legacy),
        doServerRequest(actualBaseUrl, actual ?? legacy, stores.actual),
      ])
      // Then
      validateAssertions(name, actualRes)
      if (verbose) {
        console.log('legacyRes:', legacyRes)
        console.log('actualRes:', actualRes)
      }
      const diffs = compareResponses(legacyRes, actualRes)
      expect(diffs).toStrictEqual([])
    },
  )
})
