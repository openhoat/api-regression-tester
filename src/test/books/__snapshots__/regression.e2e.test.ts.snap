// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`API regression test should return equivalent results for "GET a book given a title" request: GET a book given a title  data 1`] = `
[
  {
    "author": "Antoine de Saint-Exupéry",
    "date": "13/2/2011",
    "id": Any<String>,
    "title": "Le Petit Prince",
  },
]
`;

exports[`API regression test should return equivalent results for "GET a book given a title" request: GET a book given a title  status 1`] = `200`;

exports[`API regression test should return equivalent results for "GET a book" request: GET a book  data 1`] = `
{
  "author": "Victor Hugo",
  "date": "13/2/2019",
  "id": Any<String>,
  "title": "Notre-Dame de Paris",
}
`;

exports[`API regression test should return equivalent results for "GET a book" request: GET a book  status 1`] = `200`;

exports[`API regression test should return equivalent results for "GET books" request: GET books  data 1`] = `
[
  {
    "author": "Antoine de Saint-Exupéry",
    "date": "13/2/2011",
    "id": Any<String>,
    "title": "Le Petit Prince",
  },
  {
    "author": "Victor Hugo",
    "date": "13/2/2019",
    "id": Any<String>,
    "title": "Notre-Dame de Paris",
  },
]
`;

exports[`API regression test should return equivalent results for "GET books" request: GET books  status 1`] = `200`;
